enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,355 @@
+---
+title: ingress-node-firewall-with-ebpf
+authors:
+  - "@msherif1234"
+reviewers:
+  - "@danwinship"
+  - "@dave-tucker"
+  - "@dcbw"
+  - "@trozet"
+approvers:
+  - "@danwinship"
+  - "@knobunc"
+api-approvers:
+  - "@danwinship"
+creation-date: 2022-06-15
+last-updated: 2022-06-15
+status: implementable
+tracking-link:
+  - https://issues.redhat.com/browse/SDN-2781
+---
+
+# Ingress node Firewall using eBPF
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OCP customers would like to secure OCP's nodes from external DDoS attacks by configuring ingress firewall policies.
+eBPF provides very flexible filtering mechansim to allow early detection and filtering for a wide range of protocols IPv4/IPv6/ICMP/TCP/UDP/ICMPv6/SCTP.
+
+New cluster scoped CRD `GolbalNodeFirewall` will be used to configure node ingress firewall match rules and a per matching rule statistics will be avaliable.
+
+## Motivation
+
+Customers would like to protect OCP cluster from external DDoS attacks.
+
+### Goals
+
+- Solution based on using `eBPF` for packet filtering.
+- flexiable match filters rules for customer to use with a CRD validation to detect and fail invalid configuration.
+- Provide a mechanism for customers to select which nodes via `nodeSelector` and the node's interface(s) where the firewall policy will be attached to.
+- Ingress node firewall is deployed on a fully operational OCP cluster (i.e. cluster comes up with no default ingress firewall policy).
+- fail safe in case of user provisioned invalid rules which can lockup cluster operation.
+
+### Non-Goals
+
+## Proposal
+
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will prepare `GlobalNodeFirewall` object with all required firewall rules and deploy.
```

Comment:
Yeah, they won't like that either. Especially, they may want to have multiple objects with a `0.0.0.0/0` CIDR.

Like, one firewall object for "block access to the apiserver" that only applies to the masters, and one for "block weird ICMP packets" that applies to all nodes.

You'll probably need a `priority` field to determine the ordering between them.

(This is partially based on discussions from the `AdminNetworkPolicy` working group... we will definitely want to make this firewall object have a similar "feel" to ANP, in terms of things like `priority`.)

(Current ANP spec is https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/2091-admin-network-policy but updates are happening in https://github.com/kubernetes-sigs/network-policy-api/pull/30 as we nail down the details of the API.)

Reply:
I prefer using order field much easier from usability since its much more clearer the execution order vs using priority , also would like to introduce nodeendpoint object so we have policy on specified interface and different on others or on all so what about something like this 
- sample NodeEndPoint object
```yaml
kind: NodeEndpoint
metadata:
  name: worker-all-interfaces
  labels:
    role: k8s-worker
spec:
  interfaces: 

---

kind: NodeEndpoint
metadata:
  name: worker-eth0
  labels:
    role: k8s-worker
spec:
  interfaces: 
  - "eth0"

---
```

- sample `IngressNodeFirewall` object

```yaml
kind: IngressNodeFirewall
metadata:
  name: worker-nodes-firewall-on-eth0
  annotations:
    node-endpoint: worker-eth0
spec:
  nodeSelector: has(do-node-ingress-firewall)
  ingress:
  - fromCIDRS:
       - 1.1.1.1/24
       - 100:1::1/64
    rules:
    - order: 10  
      protocol: tcp
      port: 100
      action: allow
    - order: 20
      protocol: udp
      ports: [8000, 8080, 9000]
      action: allow
  - fromCIDRS:
       - 3.3.3.3/24
    rules:
    - order: 10
      protocol: icmp
      type: 3 # Destination unreachable
      code: 1 # Host unreachable
      action: allow
  - fromCIDRS:
       - 10:10::1/64
    rules:
    - order: 10
      protocol: icmpv6
      type: 128 # Ping request
      action: deny
  - fromCIDRS:
       - 0.0.0.0/0
       - 0::0/0
    rules:
    - order: 10
      protocol: tcp
      action: allow

---
kind: IngressNodeFirewall
metadata:
  name: worker-nodes-firewall
  annotations:
    node-endpoint: worker-all-interfaces
spec:
  nodeSelector: has(do-node-ingress-firewall)
  ingress:
  - fromCIDRS:
       - 1.1.1.1/24
       - 100:1::1/64
    rules:
    - order: 30
      protocol: udp
      ports: 8000-9000
      action: deny
  - fromCIDRS:
       - 0.0.0.0/0
       - 0::0/0
    rules:
    - order: 20
      protocol: tcp
      action: deny
    - order: 30
      protocol: udp
      action: deny
    - order: 40
      protocol: sctp
      action: deny
```
let me know your thoughts before reworking the APIs